src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,840 @@
+        // this comes into place if our own initialized connection fails
+        // but afterwards the added neighbor builds a connection to us.
+        hostsWhitelist.add(ipAddress);
+
+        // map the ip address to the domain
+        ipToDomainMapping.put(ipAddress, addr.getHostString());
+
+        // make the identity of the newly added neighbor clear, so that it gets rejected during handshaking
+        // finalisation, in case the communicated server socket port is wrong.
+        allowedNeighbors.add(String.format("%s:%d", addr.getAddress().getHostAddress(), addr.getPort()));
+
+        // init new TCP socket channel
+        SocketChannel tcpChannel = SocketChannel.open();
+        tcpChannel.socket().setTcpNoDelay(true);
+        tcpChannel.socket().setSoLinger(true, 0);
+        tcpChannel.configureBlocking(false);
+        tcpChannel.connect(addr);
+        Neighbor neighbor = new NeighborImpl<>(selector, tcpChannel, addr.getAddress().getHostAddress(), addr.getPort(),
+                txPipeline);
+        neighbor.setDomain(addr.getHostName());
+        tcpChannel.register(selector, SelectionKey.OP_CONNECT, neighbor);
+    }
+
+    /**
+     * Checks whether the given host is allowed to connect given its IP address. <br/>
+     * The connection is allowed when:
+     * <ul>
+     * <li>the IP address is not in the {@link NeighborRouter#hostsBlacklist}</li>
+     * <li>{@link BaseIotaConfig#getMaxNeighbors()} has not been reached</li>
+     * <li>is whitelisted in {@link NeighborRouter#hostsWhitelist} (if {@link BaseIotaConfig#isAutoTetheringEnabled()}
+     * is false)</li>
+     * </ul>
+     * The IP address is blacklisted to mute it from subsequent connection attempts. The blacklisting is removed if the
+     * IP address is added through {@link NeighborRouter#addNeighbor(String)}.
+     * 
+     * @param ipAddress       The IP address
+     * @param newNeighborConn The {@link SocketChannel} to close if the connection is not allowed
+     * @return true if allowed, false if not
+     * @throws IOException if anything goes wrong closing the {@link SocketChannel}
+     */
+    private boolean okToConnect(String ipAddress, SocketChannel newNeighborConn) throws IOException {
+        if (hostsBlacklist.contains(ipAddress)) {
+            // silently drop connection
+            newNeighborConn.close();
+            return false;
+        }
+        if (availableNeighborSlotsFilled()) {
+            log.info("dropping new connection from {} as all neighbor slots are filled", ipAddress);
+            newNeighborConn.close();
+            return false;
+        }
+        boolean whitelisted = hostsWhitelist.contains(ipAddress);
+        if (!whitelisted) {
+            if (!config.isAutoTetheringEnabled()) {
+                log.info("blacklisting/dropping new connection as neighbor from {} is not defined in the config",
+                        ipAddress);
+                hostsBlacklist.add(ipAddress);
```

Comment:
An attacker would need a large amount of IP addresses in order to fill up the given set and crash the node with a mem overload. This is used in order to mute the reconnects from the given blacklisted ip address, instead of spitting out logs all the time. (check `okToConnect()`). Meaning subsequent reconnect attempts by the given IP are simply dropped without logging anything.

I don't think this is an actual attack vector at all and it seems unlikely that someone would have that many IP addresses to fill up the memory.